addons/mail/static/src/new/composer/composer_hook.js

```diff
@@ -0,0 +1,161 @@
+                if (candidatePosition < 0 || candidatePosition >= content.length) {
+                    continue;
+                }
+                const candidateChar = content[candidatePosition];
+                if (candidateChar === "/" && candidatePosition !== 0) {
+                    continue;
+                }
+                if (!suggestionDelimiters.includes(candidateChar)) {
+                    continue;
+                }
+                const charBeforeCandidate = content[candidatePosition - 1];
+                if (charBeforeCandidate && !/\s/.test(charBeforeCandidate)) {
+                    continue;
+                }
+                Object.assign(self.search, {
+                    delimiter: candidateChar,
+                    position: candidatePosition,
+                    term: content.substring(candidatePosition + 1, selectionStart),
+                });
+                self.state.count++;
+                return;
+            }
+            self.clearSearch();
+        },
+        fetch: {
+            inProgress: false,
+            rpcFunction: undefined,
+        },
+        insert(option) {
+            const cursorPosition = comp.props.composer.selection.start;
+            const content = comp.props.composer.textInputContent;
+            let textLeft = content.substring(0, self.search.position + 1);
+            let textRight = content.substring(cursorPosition, content.length);
+            if (self.search.delimiter === ":") {
+                textLeft = content.substring(0, self.search.position - 1);
+                textRight = content.substring(cursorPosition, content.length);
+            }
+            const recordReplacement = option.label;
+            if (option.partner) {
+                self.rawMentions.partnerIds.add(option.partner.id);
+            }
+            self.clearSearch();
+            comp.props.composer.textInputContent = textLeft + recordReplacement + " " + textRight;
+            comp.props.composer.selection.start = textLeft.length + recordReplacement.length + 1;
+            comp.props.composer.selection.end = textLeft.length + recordReplacement.length + 1;
+            comp.props.composer.forceCursorMove = true;
+        },
+        async process(func) {
+            if (self.fetch.inProgress) {
+                self.fetch.rpcFunction = func;
+                return;
+            }
+            self.fetch.inProgress = true;
+            self.fetch.rpcFunction = undefined;
+            await func();
+            self.fetch.inProgress = false;
+            if (self.fetch.nextMentionRpcFunction) {
+                self.process(self.fetch.nextMentionRpcFunction);
+            }
+        },
+        rawMentions: {
+            partnerIds: new Set(),
+        },
+        search: {
+            delimiter: undefined,
+            position: undefined,
+            term: undefined,
+        },
+        state: useState({
+            count: 0,
+            items: [],
+        }),
```

Comment:
why there are so many sub-object? it feels like over-classification, it's hard to guess what to find where

Reply:
All was in `this` of `Composer` component.
To simplify code in hook, I put it in a single object.

I will do follow-up PR to expose only what is necessary. And also internal code can remove `self`, so should improve slightly readability of the component hook.